text/0000-glimmer-components.md

```diff
@@ -0,0 +1,835 @@
+should be those outlined in the motivation of this RFC, and the motivation for
+the changes should be included in a summary that helps users understand the
+intent of new APIs so they can use them effectively.
+
+**Outer HTML semantics** should be fairly simple to communicate. These should
+also be the easiest to codemod, since it is mostly a 1-to-1 conversion from the
+legacy component definition to the template.
+
+**Namespaced arguments** will be a sticking point for some. The main thing to
+focus on here is aliasing to provide defaults, and the benefits of immutability.
+Patterns like the `@arg` decorator in
+[sparkles-decorators](https://github.com/gossi/sparkles-decorators) make this
+fairly ergonomic, and for simpler components defaulting using a class field
+initializer can also work.
+
+**Lifecycle hooks** may be fairly confusing to existing users, who are used to
+having a variety of hooks to choose from when coordinating lifecycle events. The
+bullet points here are:
+
+* `didCreate` is `didInsertElement`, without elements
+* `didUpdate` is the same as it is in legacy components
+* `destroy` is the correct place for all teardown code
+* `didReceiveAttrs`, `willRender`, and `didRender` code should generally be
+  extracted into a separate method called by both `didCreate` and `didUpdate`.
+* Any code that may cause a rerender in `didUpdate` or `didCreate` should
+  ideally be pushed into getters or helpers, so that it naturally is pulled into
+  the render instead of imperatively updated.
+
+**Property tracking** will be the most difficult change to understand in some
+regards, and will rely on dialing in the teaching for the mechanics of tracked
+properties (discussed in the tracked properties RFC).
+
+## Drawbacks
+
+The major drawback to Glimmer components is that they add a separate API for
+components, meaning that for the forseeable future Ember users will likely
+need to learn how to use both interchangeably. This introduces a fair amount of
+mental overhead for users, but the benefits of Glimmer components and their
+simplicity should make this less problematic.
+
+## Alternatives
+
+### No owner in constructor
+
+Sparkles components do not provide access to the owner or injections in the
+constructor, though it is a requested feature. We could add a `willCreate` or
+`init` hook which allows users to setup the instance before the first render.
+
+Alternatively, the exact method by which the owner is passed to the constructor
+can be changed (on an object vs directly) or all injections could be passed,
+enabling typed injections.
+
+### More/different lifecycle hooks
+
+The proposed lifecycle hooks are significantly more minimal than in legacy
+components. This API space could always be added to in future RFCs, but it may
+be too small to start with. In particular, lacking an equivalent to `willUpdate`
+and `didUpdateAttrs` means that some users may be forced into triggering a
```

Comment:
I actually think that `didUpdateAttrs` is quite easy to implement in the current design.

```js
export default class extends Component {
  get args() {
    return this._args;
  }

  set args(args) {
    let lastArgs = this._args;
    this._args = args;

    // do whatever you need to do...
  }
}
```